### DISCUSSION TITLE
Relating 'types' to the decompiler

### DISCUSSION THREAD

**Discussion by Wall-AF** [NationalSecurityAgency/ghidra#3621] [Q&A]
The application I'm REing is compiled for 16-bit Windows and mainly uses 8-/16-bit registers of the x86 processor, the occasional 32-bit register and the four 16-bit segment registers.  The `.cspec` file defines the sizes for all base types as expected, minus something for 32-bit addresses (which I assume are supposedly taken care of by the use of the segment registers - but that is another issue`*`).  However anything using 32-bits seems to struggle to be decompiled _nicely_ with respect to type `long` and other types/enum/etc. I've defined manually.

Here's one example: to get enums to be recognised automatically in the `Decompile`view, the standard 8-byte enum doesn't get sized apprpriately so I've had to create specifically sized versions.  (I've noticed that these versions don't appear to recognise the high bit of the specific size in the same way it does for an 8 byte enum and therefore all appear as positive numbers (may not be significant).)  This method fails when a 32-bit value is set directly in the standard registers `DX`&`AX` used for all 32-bit returns from function calls in the prototypes used.  However, if an intermediate 32-bit register is used and that is then split into the `DX`&`AX`, it works!

In the example above, the final `return ...` statement ends up:
`return CONCAT22((undefined2)lRetVal._2_2_,(undefined2)lRetVal);`
instead of:
`return lRetVal;`

To me, this kind of interpretation means that somewhere, the defined type and importantly it's size ,aren't being taken into account by the decompilation process correctly as it should combine them initially.  Any suggestions on how to fix this and other similar situations (`*`one being using a functions stack based variables address (32-bit, `SS:(SP+0x??)`) in a subsequent call)?

**Comment by astrelsky** [comment]
Looks like the joined registers for the return storage are backwards.

**Reply by Wall-AF** [reply]
`DX` is the high word and `AX` the low one.

**Comment by Wall-AF** [comment]
Another example of something not quite right is this snippet:
```
      10e0:1895 66 0f bf        MOVSX      EAX,word ptr [BP + local_6]
                46 fc
      10e0:189a 66 50           PUSH       EAX
```
decompiles to
`      iVar3 = (int)local_6 >> 0xf;`

And, if I try to set the type of `iVar3` to `long`, Ghidra reports this:
![image](https://user-images.githubusercontent.com/67698927/141146777-8689dabf-92b2-474d-84dd-696105aefa92.png)

**Reply by Wall-AF** [reply]
Yet another example:
```
   1008:1369 01 46 e6        ADD        word ptr [BP + local_1c],AX
   1008:136c 11 56 e8        ADC        word ptr [BP + local_1c+0x2],DX
```
decompiles to:
```
            bVar10 = CARRY2((uint)local_1c,uVar8);
            (uint)local_1c += uVar8;
            local_1c._2_2_ += (uint)bVar10;
```
even though the variable `local_1c` is defined as `long` in both the Assembly and Decompiler panels!

Can this be fixed? @caheckman
